guardian/recipes-backend
# Bringing together the work into a single lambda function

## What does this change?

Implements https://trello.com/c/msTNa5WY.

With this, the first version of the content pipeline is complete!

### Content update

- extract recipes from article (#12 )
- find recipes to remove/update
- takedown superceded recipes
- upload new versions
- update the index if we've made any content changes

### Content update (retrievable)

- mostly already implemented in #9 
- adding tests, separating from the content-update handler to make the test mocking easier

### Content delete

- already implemented in #11 

## How to test

1. Take a snapshot of the current index state: `curl https://recipes.code.dev-guardianapis.com/index.json | jq > before.json`
2. Go into Composer CODE and create a new article with a recipe in it.  Alternatively, edit the recipe blob of an existing one.
3. Launch your changes
4. Wait for it to propagate.... 60s should be more than enough.  If you're curious, open the Content Platforms space in Central ELK; go to Discover; then open the saved search "Recipe responder"
5. The index should be updated by now.  Take a snapshot of the updated index: `curl https://recipes.code.dev-guardianapis.com/index.json | jq > after.json`
6. Run `diff -u after.json before.json`
7. You should see the timestamp in the index.json has moved and at least one recipe blob has changed.  If you don't, then reload a couple of times more - there are a couple of levels of CDN for it to get through.
8. Retrieve your recipe content from `https://recipes.code.dev-guardianapis.com/content/{checksum-value}`
9. Delete the recipe from your article
10. Re-launch
11. Repeat steps 4-7.  The content should now be gone.
12. Add a recipe again
13. Re-launch
14. Verify it's present in the index
15. Take down the article
16. Repeat steps 4-7.  The content should now be gone. 

## How can we measure success?

Basecamp are able to consume recipes published from Composer
